src/arch/aarch64/kernel/serial.rs

```diff
@@ -26,41 +59,34 @@ impl ErrorType for SerialDevice {
 
 impl Read for SerialDevice {
 	fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
-		let _ = buf;
-		Ok(0)
+		let mut guard = UART_DEVICE.lock();
+
+		if guard.buffer.is_empty() {
+			Ok(0)
+		} else {
+			let min = buf.len().min(guard.buffer.len());
+
+			for (dst, src) in buf[..min].iter_mut().zip(guard.buffer.drain(..min)) {
+				*dst = src;
+			}
```

Comment:
Why not use `copy_from_slice` here like on x86?

https://github.com/hermit-os/kernel/blob/8c012c672d5dd044d43c4b0e611b457028e14227/src/arch/x86_64/kernel/serial.rs#L55-L64

Reply:
x86 allocates an intermediate Vec, this version doesn't require an allocation

Reply:
Oh, I see.
So we should change x86 in a separate PR, right?

I was now thinking about using `VecDeque::as_slices`, but we could do even better by just forwarding all methods to the `Read` implementation for `VecDeque`, right?